From 69e5d7cf0196e91ee3d3d2083cb57faac88d8473 Mon Sep 17 00:00:00 2001
From: tonyfan <waterwet@outlook.com>
Date: Sat, 29 Aug 2026 01:20:30 +0800
Subject: [PATCH] fix: use native submenu navigation

Launch submenus through SubSettings so toolbar Up and predictive back use standard Settings activity transitions.
---
 src/com/infinity/suite/InfinitySuite.kt | 84 ++++++++++---------------
 1 file changed, 34 insertions(+), 50 deletions(-)

diff --git a/src/com/infinity/suite/InfinitySuite.kt b/src/com/infinity/suite/InfinitySuite.kt
index f4b2955..f6de131 100644
--- a/src/com/infinity/suite/InfinitySuite.kt
+++ b/src/com/infinity/suite/InfinitySuite.kt
@@ -16,26 +16,20 @@
 
 package com.infinity.suite
 
-import android.animation.AnimatorInflater
-import android.animation.ObjectAnimator
 import android.app.Activity
 import android.content.Intent
 import android.content.pm.ActivityInfo
 import android.content.res.Configuration
 import android.os.Bundle
-import android.view.animation.AnimationUtils
-import android.view.animation.Interpolator
 import android.view.LayoutInflater
 import android.view.Surface
 import android.view.View
 import android.view.ViewGroup
-import androidx.fragment.app.Fragment
-import androidx.fragment.app.FragmentManager
-import androidx.fragment.app.FragmentTransaction
 import com.android.internal.logging.nano.MetricsProto
 import com.android.internal.util.infinity.Utils
 import com.android.settings.R
 import com.android.settings.SettingsPreferenceFragment
+import com.android.settings.core.SubSettingLauncher
 import com.google.android.material.card.MaterialCardView
 import com.android.settings.WallpaperBlurView
 
@@ -91,19 +85,19 @@ class InfinitySuite : SettingsPreferenceFragment(), View.OnClickListener {
     }
 
     override fun onClick(view: View) {
-        val fragment = when (view.id) {
-            R.id.quicksettings_card -> QuickSettings()
-            R.id.monetsettings_card -> MonetSettings()
-            R.id.statusbarsettings_card -> StatusBarSettings()
-            R.id.lockscreensettings_card -> LockScreenSettings()
-            R.id.powersettings_card -> PowerMenuSettings()
-            R.id.teaminfo_card -> TeamInfo()
-            R.id.notificationsettings_card -> NotificationSettings()
-            R.id.theme_card -> ThemesSettings()
-            R.id.navigationsettings_card -> NavbarSettings()
-            R.id.miscsettings_card -> MiscSettings()
-            R.id.buttonsettings_card -> ButtonSettings()
-            R.id.aodsettings_card -> AmbientCustomizations()
+        val destination = when (view.id) {
+            R.id.quicksettings_card -> QuickSettings::class.java.name
+            R.id.monetsettings_card -> MonetSettings::class.java.name
+            R.id.statusbarsettings_card -> StatusBarSettings::class.java.name
+            R.id.lockscreensettings_card -> LockScreenSettings::class.java.name
+            R.id.powersettings_card -> PowerMenuSettings::class.java.name
+            R.id.teaminfo_card -> TeamInfo::class.java.name
+            R.id.notificationsettings_card -> NotificationSettings::class.java.name
+            R.id.theme_card -> ThemesSettings::class.java.name
+            R.id.navigationsettings_card -> NavbarSettings::class.java.name
+            R.id.miscsettings_card -> MiscSettings::class.java.name
+            R.id.buttonsettings_card -> ButtonSettings::class.java.name
+            R.id.aodsettings_card -> AmbientCustomizations::class.java.name
             R.id.customization_picker_button -> {
                 openCustomizationPickerActivity()
                 return
@@ -111,41 +105,31 @@ class InfinitySuite : SettingsPreferenceFragment(), View.OnClickListener {
             else -> null
         }
 
-        val title = when (view.id) {
-            R.id.quicksettings_card -> getString(R.string.quicksettings_title)
-            R.id.monetsettings_card -> getString(R.string.monet_title)
-            R.id.statusbarsettings_card -> getString(R.string.statusbar_title)
-            R.id.lockscreensettings_card -> getString(R.string.lockscreen_title)
-            R.id.powersettings_card -> getString(R.string.powermenu_title)
-            R.id.teaminfo_card -> getString(R.string.team_title)
-            R.id.notificationsettings_card -> getString(R.string.notifications_title)
-            R.id.theme_card -> getString(R.string.themes_title)
-            R.id.navigationsettings_card -> getString(R.string.navbar_title)
-            R.id.miscsettings_card -> getString(R.string.misc_title)
-            R.id.buttonsettings_card -> getString(R.string.button_title)
-            R.id.aodsettings_card -> getString(R.string.ambient_text_category_title)
-            else -> null
+        val titleRes = when (view.id) {
+            R.id.quicksettings_card -> R.string.quicksettings_title
+            R.id.monetsettings_card -> R.string.monet_title
+            R.id.statusbarsettings_card -> R.string.statusbar_title
+            R.id.lockscreensettings_card -> R.string.lockscreen_title
+            R.id.powersettings_card -> R.string.powermenu_title
+            R.id.teaminfo_card -> R.string.team_title
+            R.id.notificationsettings_card -> R.string.notifications_title
+            R.id.theme_card -> R.string.themes_title
+            R.id.navigationsettings_card -> R.string.navbar_title
+            R.id.miscsettings_card -> R.string.misc_title
+            R.id.buttonsettings_card -> R.string.button_title
+            R.id.aodsettings_card -> R.string.ambient_text_category_title
+            else -> 0
         }
 
-        if (fragment != null && title != null) {
-            replaceFragmentWithSystemAnimations(fragment, title)
+        if (destination != null && titleRes != 0) {
+            SubSettingLauncher(requireContext())
+                .setDestination(destination)
+                .setTitleRes(titleRes)
+                .setSourceMetricsCategory(metricsCategory)
+                .launch()
         }
     }
 
-    private fun replaceFragmentWithSystemAnimations(fragment: Fragment, title: String) {
-        val transaction = parentFragmentManager.beginTransaction()
-        
-        transaction.setCustomAnimations(
-	    R.anim.fragment_slide_in,
-	    R.anim.fragment_slide_out
-        )
-        transaction.replace(this.id, fragment)
-        transaction.addToBackStack(null)
-        transaction.commitAllowingStateLoss()
-
-        activity?.title = title
-    }
-
     override fun getMetricsCategory(): Int {
         return MetricsProto.MetricsEvent.INFINITY
     }
-- 
2.34.1
